data_layout.go

```diff
@@ -0,0 +1,421 @@
+
+type IntegerSizeAlignment struct { // All sizes are in bits.
+	size               uint64
+	abiAlignment       uint64
+	preferredAlignment uint64 // optional and defaults to abi
+}
+
+func NewIntegerSizeAlignment(size, abiAl, prefAl uint64) *IntegerSizeAlignment {
+	return &IntegerSizeAlignment{size: size, abiAlignment: abiAl, preferredAlignment: prefAl}
+}
+
+func getDefaultIntAligns() map[uint64]*IntegerSizeAlignment {
+	intAlignMap := make(map[uint64]*IntegerSizeAlignment)
+	intAlignMap[1] = NewIntegerSizeAlignment(1, 8, 8)
+	intAlignMap[8] = NewIntegerSizeAlignment(8, 8, 8)
+	intAlignMap[16] = NewIntegerSizeAlignment(16, 16, 16)
+	intAlignMap[32] = NewIntegerSizeAlignment(32, 32, 32)
+	intAlignMap[64] = NewIntegerSizeAlignment(64, 32, 64)
+	return intAlignMap
+}
+
+type VectorSizeAlignment struct { // All sizes are in bits.
+	size               uint64
+	abiAlignment       uint64
+	preferredAlignment uint64 // optional and defaults to abi
+}
+
+func NewVectorSizeAlignment(size, abiAl, prefAl uint64) *VectorSizeAlignment {
+	return &VectorSizeAlignment{size: size, abiAlignment: abiAl, preferredAlignment: prefAl}
+}
+
+func getDefaultVectorAligns() map[uint64]*VectorSizeAlignment {
+	vectorAlignMap := make(map[uint64]*VectorSizeAlignment)
+	vectorAlignMap[64] = NewVectorSizeAlignment(64, 64, 64)
+	vectorAlignMap[128] = NewVectorSizeAlignment(128, 128, 128)
+	return vectorAlignMap
+}
+
+type FloatingPointSizeAlignment struct { // All sizes are in bits.
+	size               uint64
+	abiAlignment       uint64
+	preferredAlignment uint64 // optional and defaults to abi
+}
+
+func NewFloatingPointSizeAlignment(size, abiAl, prefAl uint64) *FloatingPointSizeAlignment {
+	return &FloatingPointSizeAlignment{size: size, abiAlignment: abiAl, preferredAlignment: prefAl}
+}
+
+func getDefaultFloatAligns() map[uint64]*FloatingPointSizeAlignment {
+	floatAlignMap := make(map[uint64]*FloatingPointSizeAlignment)
+	floatAlignMap[16] = NewFloatingPointSizeAlignment(16, 16, 16)
+	floatAlignMap[32] = NewFloatingPointSizeAlignment(32, 32, 32)
+	floatAlignMap[64] = NewFloatingPointSizeAlignment(64, 64, 64)
+	floatAlignMap[128] = NewFloatingPointSizeAlignment(128, 128, 128)
+	return floatAlignMap
+}
+
+type AggregateAlignment struct {
+	abiAlignment       uint64
+	preferredAlignment uint64 // optional and defaults to abi
+}
+
+func NewAggregateAlignment(abiAl, prefAl uint64) *AggregateAlignment {
+	return &AggregateAlignment{abiAlignment: abiAl, preferredAlignment: prefAl}
+}
+
+type FunctionPointerAlignment struct {
+	isIndependant bool
+	abiAlignment  uint64
+}
+
+func NewFunctionPointerAlignment(isInd bool, abiAl uint64) *FunctionPointerAlignment {
+	return &FunctionPointerAlignment{isIndependant: isInd, abiAlignment: abiAl}
+}
+
+type ManglingStyle string
+
+const (
+	ELF        ManglingStyle = "e"
+	Mips       ManglingStyle = "m"
+	MachO      ManglingStyle = "o"
+	WinX86COFF ManglingStyle = "x"
+	WinCOFF    ManglingStyle = "w"
+	XCOFF      ManglingStyle = "a"
+)
```

Comment:
I would like to see them in a separate file.

Reply:
I think a submodule for `datalayout` is ok for me.

Reply:
I initially had data_layout_util.go file with those constants and other utils/helpers/string(parsing/generation) other than the main structure and its methods present in this file. But merged it before sending this PR.

Reply:
ok, it's fine for now.